subworkflows/local/metagenomics_profiling.nf

```diff
@@ -0,0 +1,160 @@
+            ch_input_for_malt_tmp =  reads
+                .map { meta, reads -> reads }
+                .collate( params.metagenomics_malt_group_size ) //collate into bins of defined lengths
+                .map{
+                    reads ->
+                    // add new meta with db-name as id
+                    [[id: file(params.metagenomics_profiling_database).getBaseName() ], reads]
+                }
+                .combine(database) //combine with database
+                .multiMap{
+                    // and split apart again
+                    meta, reads, database ->
+                        reads: [meta, reads]
+                        database: database
+                }
+            ch_input_for_malt = ch_input_for_malt_tmp.reads
+            database = ch_input_for_malt_tmp.database
+        }
+
+        else {
+            ch_input_for_malt =  reads
+                .map { meta, reads -> reads }
+                .collect()
+                .map{
+                    // make sure id is db_name for publishing purposes.
+                    reads ->
+                    [[id: file(params.metagenomics_profiling_database).getBaseName() ], reads]
+                }
+        }
+
+        // Run MALT
+        MALT_RUN ( ch_input_for_malt, database )
+
+        ch_maltrun_for_megan = MALT_RUN.out.rma6
+            .transpose()
+            .map {
+                meta, rma ->
+                    // re-extract meta from file names, use filename without rma to
+                    // ensure we keep paired-end information in downstream filenames
+                    // when no pair-merging
+                    def meta_new = meta.clone()
+                    meta_new['db_name'] = meta.id
+                    meta_new['id'] = rma.baseName
+                [ meta_new, rma ]
+            }
+
+        ch_versions                 = ch_versions.mix( MALT_RUN.out.versions.first() )
+        ch_raw_classifications      = ch_raw_classifications.mix( ch_maltrun_for_megan )
+        ch_multiqc_files            = ch_multiqc_files.mix( MALT_RUN.out.log )
+        ch_postprocessing_input     = ch_postprocessing_input.mix( ch_maltrun_for_megan )
+    }
+
+    else if ( params.metagenomics_profiling_tool == 'metaphlan' ) {
+
+        reads = reads.combine(database)
+        metaphlan_reads = reads.map{ meta, reads, database -> [meta, reads] }
+        metaphlan_db = reads.map{ meta, reads, database -> [database] }
+
+        METAPHLAN_METAPHLAN ( metaphlan_reads , metaphlan_db )
+        ch_versions             = ch_versions.mix( METAPHLAN_METAPHLAN.out.versions.first() )
+        ch_raw_profiles         = ch_raw_profiles.mix( METAPHLAN_METAPHLAN.out.profile )
+        ch_postprocessing_input = ch_postprocessing_input.mix( METAPHLAN_METAPHLAN.out.profile )
+
+    }
+
+    if ( params.metagenomics_profiling_tool == 'krakenuniq' ) {
+        // run kraken uniq per sample, to preserve the meta-data
```

Comment:
Is this still correct here? Running krakenuniq with preloaded mode per sample doesn't make sewnse, as the purposes of the preloaded is to load the database once (a bit like MALT).

Reply:
In my tests preloading is always _much_ slower than running individual jobs in threads. It makes sense in terms of reducing strain on the file-system though... I dont have a strong opinion about that and went for what I do in my pipelines :D

Reply:
Going for your sugestion now